### PR TITLE
Camera controller

### PR DESCRIPTION
**Pull request by ki3kia**

### CODE REVIEW

File: src/gameUtils.ts [hunk 1]
```diff
@@ -6,6 +8,35 @@ export type Point = {
...
+    return { x: -1, y: 0 };
+  }
+
+  if (sinVector > Math.sin(Math.PI / 4)) {
+    return { x: 0, y: -1 };
+  }
+
+  if (sinVector < -Math.sin(Math.PI / 4)) {
+    return { x: 0, y: 1 };
+  }
```
Comment (TheForsakenSpirit) [inline src/gameUtils.ts:37]: Convert Math operation to constant values.

File: src/Camera.tsx [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+        const stream = await navigator.mediaDevices.getUserMedia({ audio: false, video: true });
+        setMediaStream(stream);
+      };
+      setupCamera();
+      return;
+    }
+    videoRef.current.srcObject = mediaStream;
+
+    return () => mediaStream.getTracks().forEach((track) => track.stop());
+  }, [mediaStream]);
```
Comment (mrlika) [inline src/Camera.tsx:20]: Break into 2 useEffects. One responsible for creating media stream. It should have no dependencies.
Another for using media stream - assigning to videoRef and stoping. Should depend on [mediaStream]

File: src/Snake.tsx [hunk 3]
```diff
@@ -95,25 +96,37 @@ export const SnakeStates = ({ pokemonId, isPaused, onEndGame }: Props): JSX.Elem
...
-      <div key={'bad_food'} className={'food bad'} style={getElementPositionStyle(gameState.food.bad)} />
-    </div>
+        <div key={'good_food'} className={'food good'} style={getElementPositionStyle(gameState.food.good)} />
+        <div key={'bad_food'} className={'food bad'} style={getElementPositionStyle(gameState.food.bad)} />
+      </div>
+      {isUseCamera ? <CameraControl /> : null}
+      <input
+        type={'checkbox'}
+        key={'isCameraControl'}
+        id={'isCameraControl'}
```
Comment (mrlika) [inline src/Snake.tsx:126]: Please configure your editor to not add unnecessary curly brackets

**Review by mrlika — CHANGES_REQUESTED** [review]

File: src/Camera.tsx [hunk 4]
```diff
@@ -1,23 +1,61 @@
...
-      return;
-    }
-    videoRef.current.srcObject = mediaStream;
+    const initCamera = async () => {
+      const stream = await navigator.mediaDevices.getUserMedia({ audio: false, video: true });
+      setMediaStream(stream);
+    };
+
+    if (!mediaStream) initCamera();
+  });
```
Comment (mrlika) [inline src/Camera.tsx:34]:
```suggestion
    initCamera();
  }, []);
```

File: src/App.tsx [hunk 5]
```diff
@@ -22,18 +23,34 @@ function App(): JSX.Element {
     setIsPaused((prev) => !prev);
   };
 
+  const onCameraControl = (direction: Point) => {
```
Comment (mrlika) [inline src/App.tsx:26]: We usually name it like
```suggestion
  const handleCameraDirection = (direction: Point) => {
```

File: src/Camera.tsx [hunk 6]
```diff
@@ -1,23 +1,61 @@
...
-export const CameraControl = (): ReactElement => {
+const model = handPoseDetection.SupportedModels.MediaPipeHands;
+const detectorConfig: handPoseDetection.MediaPipeHandsMediaPipeModelConfig = {
+  runtime: 'mediapipe',
+  solutionPath: 'https://cdn.jsdelivr.net/npm/@mediapipe/hands',
+  modelType: 'full',
+  maxHands: 1,
+};
+
+const detectorPromise = handPoseDetection.createDetector(model, detectorConfig);
```
Comment (mrlika) [inline src/Camera.tsx:17]: Each component should have its own detector e.g. to play two games on a single page simultaneously.

**Review by mrlika — CHANGES_REQUESTED** [review]

File: src/Camera.tsx (hunk 6 above)
Comment (ki3kia) [reply src/Camera.tsx:17]: Placing a detector in a component will abort the detector

File: src/App.tsx [hunk 7]
```diff
@@ -40,19 +61,26 @@ function App(): JSX.Element {
...
             setIsOnGame(false);
           }}
+          cameraDirection={cameraDirection}
+        />
+        {isUseCamera ? <CameraControl onChangeDirection={handleCameraDirection} /> : undefined}
+        <input
+          type='checkbox'
+          key='isCameraControl'
+          id='isCameraControl'
+          onChange={() => setIsUseCamera((prev) => !prev)}
```
Comment (TheForsakenSpirit) [inline src/App.tsx:71]: No need to use updater here. Just `setIsUseCamera(!isUseCamera)`

File: src/Snake.tsx [hunk 8]
```diff
@@ -45,6 +46,10 @@ export const SnakeStates = ({ pokemonId, isPaused, onEndGame }: Props): JSX.Elem
     };
   };
 
+  useEffect(() => {
+    if (!isPaused && cameraDirection) directionRef.current = cameraDirection;
+  }, [isPaused, cameraDirection]);
```
Comment (TheForsakenSpirit) [inline src/Snake.tsx:51]:
```suggestion
  directionRef.current = cameraDirection;
```

**Review by TheForsakenSpirit — CHANGES_REQUESTED** [review]

**Review by mrlika — CHANGES_REQUESTED** [review]

File: src/Camera.tsx [hunk 9]
```diff
@@ -6,56 +6,68 @@ import '@mediapipe/hands';
...
-      const detector = await detectorPromise;
       if (!videoRef.current) return;
-      const hands = await detector.estimateHands(videoRef.current, { staticImageMode: true });
+
+      const hands = await detector.estimateHands(videoRef.current, { staticImageMode: false });
       let direction: Point | undefined;
       if (hands[0]) direction = getDirectionByHandsPose(hands[0]);
 
-      if (direction) onChangeDirection(direction);
+      if (direction) onChangeDirectionRef.current(direction);
```
Comment (mrlika) [inline src/Camera.tsx:64]: Please add code to prevent execution of the 2 `detector.estimateHands` simultaneously on slow devices.

**Review by mrlika — CHANGES_REQUESTED** [review]

File: src/Camera.tsx [hunk 10]
```diff
@@ -54,18 +55,19 @@ export const CameraControl = ({ onChangeDirection }: CameraControlProps): ReactE
...
+    const getDirection = async () => {
       if (!videoRef.current) return;
-
       const hands = await detector.estimateHands(videoRef.current, { staticImageMode: false });
       let direction: Point | undefined;
       if (hands[0]) direction = getDirectionByHandsPose(hands[0]);
 
       if (direction) onChangeDirectionRef.current(direction);
-    }, 200);
+      if (hands) setTimeout(getDirection, DELAY);
```
Comment (mrlika) [inline src/Camera.tsx:65]: 1. Why AI loop should be stopped if there are no hands?
2. Use closure scope to save `timeoutId` and clear it in the cleanup.

**Review by mrlika — APPROVED** [review]

**Review by TheForsakenSpirit — APPROVED** [review]

**Pull request merged**